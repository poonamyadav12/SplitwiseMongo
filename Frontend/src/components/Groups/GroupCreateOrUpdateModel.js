import axios from 'axios';
import _ from 'lodash';
import React, { useState } from 'react';
import { Button, Card, Col, Container, Form, Modal, Row } from 'react-bootstrap';
import { connect } from 'react-redux';
import '../../App.css';
import { alertActions } from '../../_actions';
import { SERVER_URL } from '../../_constants';
import { getDefaultGroupImage } from '../../_constants/avatar';
import { AlertMessages } from '../Alert/Alert';
import { UploadImage } from '../Image/UploadImage';
import { GroupMemberList } from './GroupMemberList';

function GroupCreateOrUpdateModal(props) {
  const [name, setName] = useState(props && props.group && props.group.name || "");

  const [errorMsg, setErrorMsg] = useState([]);

  const [members, setMembers] = useState(props && props.group && props.group.members || []);

  const [imageUrl, setImageUrl] = useState(props && props.group && props.group.avatar || getDefaultGroupImage());

  function handleGroupNameChange(e) {
    e.preventDefault();
    setName(e.target.value);
  }

  function handleGroupMemberChange(members) {
    setMembers(members);
  }

  async function handleSubmit(e) {
    e.preventDefault();
    let group = {
      creator: props.group && props.group.creator || props.user.email,
      name,
      members: members.map((m) => ({ first_name: m.first_name, last_name: m.last_name, email: m.email, group_join_status: m.group_join_status })),
      avatar: imageUrl,
    };

    if (props && props.group && props.group.id) {
      group = { ...group, id: props.group.id };
    }
    try {
      const response = await axios.post(SERVER_URL + (props.createMode ? '/group/create' : '/group/update'), { group });
      props.reloadHomeView();
      props.closeModal();

    } catch (error) {
      const data = error.response.data;
      const msg = Array.isArray(data) ? data.map(d => d.message) : ["Some error occured, please try again."];
      setErrorMsg(msg);
    }
  }

  return <Modal
    show={props.isOpen}
    onHide={props.closeModal}
    keyboard={false}
    backdrop="static"
    className="group-create-modal"
    animation={false}
    style={{ width: "100vw" }}
  >
    <Modal.Header closeButton>
      <Modal.Title>{props.group ? 'UPDATE THE GROUP' : 'START A NEW GROUP'}</Modal.Title>
      <AlertMessages type="danger" messages={errorMsg} />
    </Modal.Header>
    <Modal.Body style={{ width: "60vw" }} >
      <Container fluid={true}>
        <Row>
          <Col lg={3}>
            <UploadImage defaultImage={props?.group?.avatar || getDefaultGroupImage()} label={'Group avatar'} onChange={setImageUrl}></UploadImage>
          </Col>
          <Col lg={4}><Form.Group controlId="formGroupName">
            <Form.Label>My group shall be called...</Form.Label>
            <Form.Control type="text" style={{ fontSize: '24px', width: '40rem' }} defaultValue={props?.group?.name} placeholder="1600 Pennsylvania Ave" onChange={handleGroupNameChange} />
          </Form.Group>
            <Card.Title>GROUP MEMBERS</Card.Title>
            <GroupMemberList initialMembers={props?.group?.members || [{
              first_name: props.user.first_name,
              email: props.user.email,
              last_name: props.user.last_name,
            },]} onChange={(members) => handleGroupMemberChange(members)} />
          </Col>
        </Row>
      </Container>
    </Modal.Body>
    <Modal.Footer>
      <Button variant="secondary" onClick={props.closeModal}>
        Cancel
          </Button>
      <Button variant="success" onClick={handleSubmit}>Save</Button>
    </Modal.Footer>
  </Modal>;
}

function mapState(state) {
  const { user } = state.authentication;
  return { user };
}

const actionCreators = {
  errorAlert: alertActions.error,
  clearAlert: alertActions.clear,
};

const connectedGroupCreateOrUpdateModal = connect(mapState, actionCreators)(GroupCreateOrUpdateModal);
export { connectedGroupCreateOrUpdateModal as GroupCreateOrUpdateModal, GroupCreateOrUpdateModal as GroupCreateOrUpdateModalForTest };
