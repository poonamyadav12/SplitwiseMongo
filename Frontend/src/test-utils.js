// test-utils.js
import React from 'react'
import { render as rtlRender } from '@testing-library/react'
import { Provider } from 'react-redux'
// Import your own reducer
import rootReducer from '../src/reducers';
import { createLogger } from 'redux-logger';
import { createStore, applyMiddleware, compose } from 'redux';
import thunkMiddleware from 'redux-thunk';

jest.mock('react-router-dom'); // <- Use this line
jest.mock('react-avatar-editor');
jest.mock('axios');

const loggerMiddleware = createLogger();
const composeEnhancers = (typeof window !== 'undefined' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

function render(
  ui,
  {
    initialState,
    store = createStore(
      rootReducer,
      composeEnhancers(
        applyMiddleware(
          thunkMiddleware,
          loggerMiddleware
        ))
    ),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// re-export everything
export * from '@testing-library/react'
// override render method
export { render }