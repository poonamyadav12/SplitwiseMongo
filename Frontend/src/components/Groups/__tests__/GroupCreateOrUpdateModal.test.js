import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import mockAxios from "axios";
import { setupServer } from 'msw/node';
import renderer from 'react-test-renderer'
import React from 'react';
import { fireEvent, render, screen } from '../../../test-utils';
import { GroupCreateOrUpdateModalForTest } from '../GroupCreateOrUpdateModel';
import ReactDOM from 'react-dom';

const server = setupServer()

beforeAll(() => {

  server.listen()
})
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

const user1 = {
  "avatar": "https://s3.amazonaws.com/splitwise/uploads/user/default_avatars/avatar-ruby47-100px.png",
  "default_currency": "USD",
  "email": "michael@gmail.com",
  "first_name": "Michael",
  "last_name": "Leland",
  "password": "Abc@1234",
  "registration_status": "JOINED",
  "time_zone": "Europe/Berlin"
};
const user2 = {
  "avatar": "https://s3.amazonaws.com/splitwise/uploads/user/default_avatars/avatar-ruby47-100px.png",
  "default_currency": "USD",
  "email": "brett@gmail.com",
  "first_name": "Brett",
  "last_name": "Lee",
  "password": "Abc@1234",
  "registration_status": "JOINED",
  "time_zone": "Europe/Berlin"
};

const group = {
  id: 'b839cd93-ce8b-4443-8cb6-eaa4d5ae519b',
  name: 'Farewell Party',
  avatar:
    'https://splitwise-images.s3.us-east-2.amazonaws.com/images/1616208501244',
  creator: 'michael@gmail.com',
  members: [user1, user2],
  group_join_status: ['JOINED', 'JOINED'],
};


test('displays group update modal page and makes API call', async () => {
  mockAxios.post.mockResolvedValue({
    data: group
  });

  render(<GroupCreateOrUpdateModalForTest user={user1} group={group} isOpen={true} />)

  const groupName = screen.getByPlaceholderText('1600 Pennsylvania Ave');
  expect(groupName).toHaveValue('Farewell Party');
  fireEvent.change(groupName, { target: { value: 'Farewell party again' } });
  expect(screen.getByText('brett@gmail.com')).toBeInTheDocument();
  expect(screen.getByText('michael@gmail.com')).toBeInTheDocument();

  fireEvent.click(screen.getByText('Save'));

  setTimeout(() => {
    expect(mockAxios.post).toBeCalledWith("http://localhost:3001/group/update", {
      group: {
        id: 'b839cd93-ce8b-4443-8cb6-eaa4d5ae519b',
        name: 'Farewell party again',
        avatar:
          'https://splitwise-images.s3.us-east-2.amazonaws.com/images/1616208501244',
        creator: 'michael@gmail.com',
        members: [user1, user2],
        group_join_status: ['JOINED', 'JOINED'],
      }
    });
  }, 100);
});


test('displays group create modal page and makes API call', async () => {
  mockAxios.post.mockResolvedValue({
    data: group
  });

  const newGroup = group;
  delete newGroup.id;
  render(<GroupCreateOrUpdateModalForTest createMode={true} user={user1} group={newGroup} isOpen={true} />)

  const groupName = screen.getByPlaceholderText('1600 Pennsylvania Ave');
  expect(groupName).toHaveValue('Farewell Party');
  fireEvent.change(groupName, { target: { value: 'Farewell party again' } });
  expect(screen.getByText('brett@gmail.com')).toBeInTheDocument();
  expect(screen.getByText('michael@gmail.com')).toBeInTheDocument();

  fireEvent.click(screen.getByText('Save'));

  setTimeout(() => {
    expect(mockAxios.post).toBeCalledWith("http://localhost:3001/group/create", {
      group: {
        name: 'Farewell party again',
        avatar:
          'https://splitwise-images.s3.us-east-2.amazonaws.com/images/1616208501244',
        creator: 'michael@gmail.com',
        members: [user1, user2],
        group_join_status: ['JOINED', 'JOINED'],
      }
    });
  }, 100);
});


test('displays group create modal page and user cancels the changes', async () => {
  mockAxios.post.mockResolvedValue({
    data: group
  });

  const newGroup = group;
  delete newGroup.id;
  render(<GroupCreateOrUpdateModalForTest createMode={true} user={user1} group={newGroup} isOpen={true} />)

  const groupName = screen.getByPlaceholderText('1600 Pennsylvania Ave');
  expect(groupName).toHaveValue('Farewell Party');
  fireEvent.change(groupName, { target: { value: 'Farewell party again' } });
  expect(screen.getByText('brett@gmail.com')).toBeInTheDocument();
  expect(screen.getByText('michael@gmail.com')).toBeInTheDocument();

  fireEvent.click(screen.getByText('Cancel'));

  setTimeout(() => {
    expect(mockAxios.post).toHaveBeenCalledTimes(0);
  }, 100);
});

test('renders correctly', () => {
  ReactDOM.createPortal = jest.fn((element, node) => {
    return element
  });
  const tree = renderer
    .create(<GroupCreateOrUpdateModalForTest user={user1} group={group} isOpen={true} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});