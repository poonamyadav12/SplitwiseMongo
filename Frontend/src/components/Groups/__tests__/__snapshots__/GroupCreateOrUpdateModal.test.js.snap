// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div
    className="modal-backdrop show"
    onClick={[Function]}
  />,
  <div
    aria-modal={true}
    className="group-create-modal modal"
    onClick={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={
      Object {
        "display": "block",
        "width": "100vw",
      }
    }
    tabIndex={-1}
  >
    <div
      className="modal-dialog"
      onMouseDown={[Function]}
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <div
            className="modal-title h4"
          >
            UPDATE THE GROUP
          </div>
          <button
            className="close"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-hidden="true"
            >
              Ã—
            </span>
            <span
              className="sr-only"
            >
              Close
            </span>
          </button>
        </div>
        <div
          className="modal-body"
          style={
            Object {
              "width": "60vw",
            }
          }
        >
          <div
            className="container-fluid"
          >
            <div
              className="row"
            >
              <div
                className="col-lg-3"
              >
                <div
                  className="col"
                  style={
                    Object {
                      "textAlign": "center",
                    }
                  }
                >
                  <div
                    className="form-group"
                  >
                    <label
                      className="form-label"
                      htmlFor="imageUplaod"
                    >
                      <h5>
                        Group avatar
                      </h5>
                    </label>
                    <div
                      onBlur={[Function]}
                      onClick={[Function]}
                      onDragEnter={[Function]}
                      onDragLeave={[Function]}
                      onDragOver={[Function]}
                      onDrop={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      tabIndex={0}
                    >
                      <input
                        autoComplete="off"
                        multiple={true}
                        onChange={[Function]}
                        onClick={[Function]}
                        style={
                          Object {
                            "display": "none",
                          }
                        }
                        tabIndex={-1}
                        type="file"
                      />
                    </div>
                    <h5>
                      Drop images or click to upload
                    </h5>
                  </div>
                </div>
              </div>
              <div
                className="col-lg-4"
              >
                <div
                  className="form-group"
                >
                  <label
                    className="form-label"
                    htmlFor="formGroupName"
                  >
                    My group shall be called...
                  </label>
                  <input
                    className="form-control"
                    defaultValue="Farewell Party"
                    id="formGroupName"
                    onChange={[Function]}
                    placeholder="1600 Pennsylvania Ave"
                    style={
                      Object {
                        "fontSize": "24px",
                        "width": "40rem",
                      }
                    }
                    type="text"
                  />
                </div>
                <div
                  className="card-title h5"
                >
                  GROUP MEMBERS
                </div>
                <div
                  className="container"
                >
                  <div
                    className="row"
                  >
                    <div
                      className="col-sm-2"
                    >
                      <div
                        className="form-group"
                      >
                        <label
                          className="form-label"
                          htmlFor="formName0"
                        >
                          <h5>
                            Michael
                             
                            Leland
                          </h5>
                        </label>
                      </div>
                    </div>
                    <div
                      className="col-sm-2"
                    >
                      <div
                        className="form-group"
                      >
                        <label
                          className="form-label"
                          htmlFor="formEmail0"
                        >
                          <h5>
                            michael@gmail.com
                          </h5>
                        </label>
                      </div>
                    </div>
                    <div
                      className="col-sm-1"
                      style={
                        Object {
                          "textAlign": "center",
                        }
                      }
                    />
                  </div>
                </div>
                <div
                  className="container"
                >
                  <div
                    className="row"
                  >
                    <div
                      className="col-sm-2"
                    >
                      <div
                        className="form-group"
                      >
                        <label
                          className="form-label"
                          htmlFor="formName1"
                        >
                          <h5>
                            Brett
                             
                            Lee
                          </h5>
                        </label>
                      </div>
                    </div>
                    <div
                      className="col-sm-2"
                    >
                      <div
                        className="form-group"
                      >
                        <label
                          className="form-label"
                          htmlFor="formEmail1"
                        >
                          <h5>
                            brett@gmail.com
                          </h5>
                        </label>
                      </div>
                    </div>
                    <div
                      className="col-sm-1"
                      style={
                        Object {
                          "textAlign": "center",
                        }
                      }
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn btn-secondary"
            disabled={false}
            type="button"
          >
            Cancel
          </button>
          <button
            className="btn btn-success"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Save
          </button>
        </div>
      </div>
    </div>
  </div>,
]
`;
